lib/archive/unpack.go

```diff
@@ -34,37 +34,31 @@ const (
 	fileWriteFlags = os.O_CREATE | os.O_TRUNC | os.O_WRONLY
 )
 
-// unpack will unpack the given tarstream(if it is a tar stream) on the local filesystem based on the unpackPath in the path spec
+// unpack will unpack the given tarstream(if it is a tar stream) on the local filesystem based on the specified root
+// combined with any rebase from the path spec
 //
 // the pathSpec will include the following elements
 // - include : any tar entry that has a path below(after stripping) the include path will be written
 // - strip : The strip string will indicate the
 // - exlude : marks paths that are to be excluded from the write operation
-// - rebase : marks the the write path that will be tacked onto the "unpackPath". e.g /tmp/unpack + /my/target/path = /tmp/unpack/my/target/path
-func Unpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, unpackPath string) error {
-	// the tar stream should be wrapped up at the end of this call
-	tr := tar.NewReader(tarStream)
-
-	strip := filter.StripPath
-	target := filter.RebasePath
+// - rebase : marks the the write path that will be tacked onto the "root". e.g /tmp/unpack + /my/target/path = /tmp/unpack/my/target/path
+func Unpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string) error {
+	op.Debugf("unpacking archive to root: %s, filter: %+v", root, filter)
 
-	if target == "" {
-		op.Debugf("Bad target path in FilterSpec (%#v)", filter)
-		return fmt.Errorf("Invalid write target specified")
-	}
-
-	if strip == "" {
-		op.Debugf("Strip path was set to \"\"")
-	}
+	tr := tar.NewReader(tarStream)
 
-	if _, err := os.Stat(unpackPath); err != nil {
+	fi, err := os.Stat(root)
+	if err != nil {
 		// the target unpack path does not exist. We should not get here.
-		op.Errorf("tar unpack target does not exist (%s)", unpackPath)
+		op.Errorf("tar unpack target does not exist: %s", root)
 		return err
 	}
 
-	finalTargetPath := filepath.Join(unpackPath, target)
-	op.Debugf("finalized target path for Tar unpack operation at (%s)", finalTargetPath)
+	if !fi.IsDir() {
```

Comment:
in this case you are actually targeting a file source to a file destination. So you are requesting that the file in the container be overwritten. I am testing this now, docker may actually reject if the file exists, not 100% sure on that. But otherwise this is definitely a use case for cp.